lib/plugins/processor/post.js

```diff
@@ -131,18 +134,19 @@ module.exports = ctx => {
 
       // FIXME: Data may be inserted when reading files. Load it again to prevent
       // race condition. We have to solve this in warehouse.
-      const doc = Post.findOne({source: file.path});
-
-      if (doc) {
-        return doc.replace(data);
-      }
-
-      return Post.insert(data);
-    }).then(doc => Promise.all([
-      doc.setCategories(categories),
-      doc.setTags(tags),
-      scanAssetDir(doc)
-    ]));
+      return Promise.resolve()
+        .then(() => {
+          const doc = Post.findOne({source: file.path});
+          if (doc) {
+            return doc.replace(data);
+          }
+          return Post.insert(data);
+        }).then(doc => Promise.all([
+          doc.setCategories(categories),
+          doc.setTags(tags),
+          scanAssetDir(doc)
+        ]));
+    });
```

Comment:
I see no necessary of changes here.

Reply:
@SukkaW I had to make those changes to break out of promise chain. Otherwise, after 

```js
      if (data.private) {
        return;
      }
```
we would go to
```js
.then(doc => Promise.all([
      doc.setCategories(categories),
      doc.setTags(tags),
      scanAssetDir(doc)
    ]));
```
doc would be undefined and everything would break apart.

Reply:
@jehy I see. It makes sense.
```suggestion
      return Promise.resolve(Post.findOne({source: file.path}))
        .then(doc => {
          if (doc) return doc.replace(data);

          return Post.insert(data);
        }).then(doc => Promise.all([
          doc.setCategories(categories),
          doc.setTags(tags),
          scanAssetDir(doc)
        ]));
    });
```

It might be possible to reduce a tick during `Promise` creation.

Reply:
I suppose that won't make impact since the number of ticks doesn't change. But it may confuse future developers - they may think that `Post.findOne` was put into `Promise.resolve` because that's an async function.